eth/catalyst/simulated_beacon.go

```diff
@@ -0,0 +1,238 @@
+		}
+	}
+}
+
+type SimulatedBeacon struct {
+	shutdownCh  chan struct{}
+	eth         *eth.Ethereum
+	period      uint64
+	withdrawals withdrawalQueue
+
+	feeRecipient     common.Address
+	feeRecipientLock sync.Mutex // lock gates concurrent access to the feeRecipient
+
+	engineAPI          *ConsensusAPI
+	curForkchoiceState engine.ForkchoiceStateV1
+	buildWaitTime      time.Duration
+	lastBlockTime      uint64
+}
+
+func NewSimulatedBeacon(period uint64, eth *eth.Ethereum) (*SimulatedBeacon, error) {
+	chainConfig := eth.APIBackend.ChainConfig()
+	if !chainConfig.IsDevMode {
+		return nil, errors.New("incompatible pre-existing chain configuration")
+	}
+	header := eth.BlockChain().CurrentHeader()
+	current := engine.ForkchoiceStateV1{
+		HeadBlockHash:      header.Hash(),
+		SafeBlockHash:      header.Hash(),
+		FinalizedBlockHash: header.Hash(),
+	}
+	engineAPI := NewConsensusAPI(eth)
+
+	// if genesis block, send forkchoiceUpdated to trigger transition to PoS
+	if header.Number.Sign() == 0 {
+		if _, err := engineAPI.ForkchoiceUpdatedV2(current, nil); err != nil {
+			return nil, err
+		}
+	}
+	return &SimulatedBeacon{
+		eth:                eth,
+		period:             period,
+		shutdownCh:         make(chan struct{}),
+		buildWaitTime:      time.Millisecond * 100,
+		engineAPI:          engineAPI,
+		lastBlockTime:      header.Time,
+		curForkchoiceState: current,
+		withdrawals:        withdrawalQueue{make(chan *types.Withdrawal, 20)},
+	}, nil
+}
+
+func (c *SimulatedBeacon) setFeeRecipient(feeRecipient common.Address) {
+	c.feeRecipientLock.Lock()
+	c.feeRecipient = feeRecipient
+	c.feeRecipientLock.Unlock()
+}
+
+// Start invokes the SimulatedBeacon life-cycle function in a goroutine
+func (c *SimulatedBeacon) Start() error {
+	if c.period == 0 {
+		go c.loopOnDemand()
+	} else {
+		go c.loop()
+	}
+	return nil
+}
+
+// Stop halts the SimulatedBeacon service
+func (c *SimulatedBeacon) Stop() error {
+	close(c.shutdownCh)
+	return nil
+}
+
+// sealBlock initiates payload building for a new block and creates a new block with the completed payload
+func (c *SimulatedBeacon) sealBlock(withdrawals []*types.Withdrawal) error {
+	tstamp := uint64(time.Now().Unix())
+	if tstamp <= c.lastBlockTime {
+		tstamp = c.lastBlockTime + 1
+	}
+	c.feeRecipientLock.Lock()
+	feeRecipient := c.feeRecipient
+	c.feeRecipientLock.Unlock()
+
+	fcResponse, err := c.engineAPI.ForkchoiceUpdatedV2(c.curForkchoiceState, &engine.PayloadAttributes{
+		Timestamp:             tstamp,
+		SuggestedFeeRecipient: feeRecipient,
+		Withdrawals:           withdrawals,
+	})
+	if err != nil {
+		return fmt.Errorf("error calling forkchoice update: %v", err)
+	}
+
+	envelope, err := c.engineAPI.GetFullPayload(*fcResponse.PayloadID)
+	if err != nil {
+		return fmt.Errorf("error retrieving payload: %v", err)
+	}
+
+	// TODO: this is a bit of a hack.  need to modify payload builder to stop payload w/ GetFullPayload
+	_, _ = c.engineAPI.GetPayloadV1(*fcResponse.PayloadID)
```

Comment:
If I don't explicitly stop building the payload with this call here, it will continue building (even after it is included in the chain).  So maybe this should be handled within the payload building code in the miner instead of a hack call here.  Will look into it.